src/marker.rs

```diff
@@ -528,21 +286,133 @@ impl<'py> Python<'py> {
     /// }
     /// ```
     ///
+    /// # Example: The `send_wrapper` loophole is closed by running the closure on dedicated thread.
+    ///
+    /// ```should_panic
+    /// # use pyo3::prelude::*;
+    /// # use pyo3::types::PyString;
+    /// use send_wrapper::SendWrapper;
+    ///
+    /// Python::with_gil(|py| {
+    ///     let string = PyString::new(py, "foo");
+    ///
+    ///     let wrapped = SendWrapper::new(string);
+    ///
+    ///     py.allow_threads(|| {
+    ///         let sneaky: &PyString = *wrapped;
+    ///
+    ///         println!("{:?}", sneaky);
+    ///     });
+    /// });
+    /// ```
+    ///
     /// [`Py`]: crate::Py
     /// [`PyString`]: crate::types::PyString
     /// [auto-traits]: https://doc.rust-lang.org/nightly/unstable-book/language-features/auto-traits.html
     /// [Parallelism]: https://pyo3.rs/main/parallelism.html
     pub fn allow_threads<T, F>(self, f: F) -> T
     where
-        F: Ungil + FnOnce() -> T,
-        T: Ungil,
+        F: Send + FnOnce() -> T,
+        T: Send,
     {
+        use std::mem::transmute;
+        use std::panic::{catch_unwind, resume_unwind, AssertUnwindSafe};
+        use std::sync::mpsc::{sync_channel, SendError, SyncSender};
+        use std::thread::{spawn, Result};
+        use std::time::Duration;
+
+        use parking_lot::{const_mutex, Mutex};
+
+        use crate::impl_::panic::PanicTrap;
+
         // Use a guard pattern to handle reacquiring the GIL,
         // so that the GIL will be reacquired even if `f` panics.
         // The `Send` bound on the closure prevents the user from
         // transferring the `Python` token into the closure.
         let _guard = unsafe { SuspendGIL::new() };
-        f()
+
+        // To close soundness loopholes w.r.t. `send_wrapper` or `scoped-tls`,
+        // we run the closure on a separate thread so that it cannot
+        // access thread-local storage from the current thread.
+
+        // 1. Construct a task
+        struct Task(*mut dyn FnMut());
+        unsafe impl Send for Task {}
+
+        let (result_sender, result_receiver) = sync_channel::<Result<T>>(0);
+
+        let mut f = Some(f);
+
+        let mut task = || {
+            let f = f.take().unwrap();
+
+            let result = catch_unwind(AssertUnwindSafe(f));
+
+            result_sender.send(result).unwrap();
```

Comment:
Does the `.unwrap()` here give us any benefit? The thread dies either way, and presumably the unwind here isn't interesting.

```suggestion
            let _ = result_sender.send(result);
```

Reply:
As written above, I wanted the make the possible failure points more explicit and a panicking thread with a detached join handle does at least trigger a log message, c.f. <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=de70d0560a08094a5527502334e014ac>

(I did consider naming the threads to generally improve usability here by indicating that these are runtime threads used by PyO3.)

Reply:
😅 great reference in the playground!

Sure thing, makes sense to keep this then 👍